res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+
+                if (allowFocusWhenParametersHidden) {
+                    engine.setValue(this.group, "show_focus", 1);
+                } else {
+                    // Connect a callback to show_parameters changing instead of
+                    // setting show_focus when effectFocusButton is pressed so
+                    // show_focus is always in the correct state, even if the user
+                    // presses the skin button for show_parameters.
+                    this.showParametersConnection = engine.makeConnection(this.group,
+                        "show_parameters",
+                        this.onShowParametersChange.bind(this));
+                    this.showParametersConnection.trigger();
+                }
+                engine.setValue(this.group, "controller_input_active", 1);
+
+                // Do not enable soft takeover upon EffectUnit construction
+                // so initial values can be loaded from knobs.
+                if (this.hasInitialized === true) {
+                    for (let n = 1; n <= 3; n++) {
+                        const effect = "[EffectRack1_EffectUnit" + this.currentUnitNumber +
+                                    "_Effect" + n + "]";
+                        engine.softTakeover(effect, "meta", true);
+                        engine.softTakeover(effect, "parameter1", true);
+                        engine.softTakeover(effect, "parameter2", true);
+                        engine.softTakeover(effect, "parameter3", true);
+                    }
+                }
+
+                this.reconnectComponents(function(component) {
+                    // update [EffectRack1_EffectUnitX] groups
+                    const unitMatch = component.group.match(script.effectUnitRegEx);
+                    if (unitMatch !== null) {
+                        component.group = eu.group;
+                    } else {
+                        // update [EffectRack1_EffectUnitX_EffectY] groups
+                        const effectMatch = component.group.match(script.individualEffectRegEx);
+                        if (effectMatch !== null) {
+                            component.group = "[EffectRack1_EffectUnit" +
+                                            eu.currentUnitNumber +
+                                            "_Effect" + effectMatch[2] + "]";
+                        }
+                    }
+                });
+            };
+
+            this.toggle = function() {
+                // cycle through unitNumbers array
+                let index = this.unitNumbers.indexOf(this.currentUnitNumber);
+                if (index === (this.unitNumbers.length - 1)) {
+                    index = 0;
+                } else {
+                    index += 1;
+                }
+                this.setCurrentUnit(this.unitNumbers[index]);
+            };
+
+            if (unitNumbers !== undefined) {
+                if (Array.isArray(unitNumbers)) {
```

Comment:
more type deduction. Any way we can avoid?

Reply:
Same reasoning in favor of it as above